### PR TITLE
Disable dashboard sync with constant. Closes #701

### LINKED ISSUE #701: Disable dashboard sync with constant
This would be helpful for sites we manage where clicking the dashboard sync problem would cause major problems (deleting the index).

### PR DESCRIPTION
**Pull request by rveitch**
## Disable Dashboard Sync

Dashboard sync can be disabled by defining the constant `EP_DASHBOARD_SYNC` as `false` in your wp-config.php file.

```php
define( 'EP_DASHBOARD_SYNC', false );
```

* This can be helpful for managed sites where users initiating a sync from the dashboard could potentially cause issues such as deleting the index and limiting this control to WP-CLI is preferred.
* When disabled, features that would require reindexing are also prevented from being enabled/disabled from the dashboard.

## Additional Details
When `EP_DASHBOARD_SYNC` is defined as `false`:
* Attempting to manually trigger a dashboard sync via `admin.php?page=elasticpress&do_sync` will not initiate a sync. This is to prevent any tomfoolery, intentional or otherwise.
* The `resume`, `pause` and `stop` buttons are not disabled by this constant. This is to prevent situations where a dashboard sync my be in progress (or paused) when sync is disabled and will allow the in-progress sync to be managed.  If no sync is in-progress, these buttons will not appear.
* Dashboard notifications for `no-sync`, `upgrade-sync` and `auto-activate-sync` will be replaced with new dismissible notifications tailored to their respective conditions.
  * These notifications are unlikely to display on the dashboard as most use cases for disabling dashboard sync should (in theory) already be handling their respective situations from WP-CLI and their trigger conditions should already be cleared, but I thought it was safer to include them for situations such as someone updating enabling a feature but forgetting to reindex.

### New Notifications
`sync-disabled-no-sync`
image

`sync-disabled-upgrade`
screenshot5028

`sync-disabled-auto-activate`
image

### CODE REVIEW

**Comment by rveitch** [comment]
I'm fairly certain that the failed checks will be due to changes in the notification strings, and I believe the tests would need to be updated to reflect new conditions introduced here.

File: classes/class-ep-feature.php [hunk 1]
```diff
@@ -162,6 +162,12 @@ public function requirements_status() {
 			$status = call_user_func( $this->requirements_status_cb, $status, $this );
 		}
 
+		if ( true === $this->requires_install_reindex && defined( 'EP_DASHBOARD_SYNC' ) && ! EP_DASHBOARD_SYNC ) {
+      $status->code = 2;
+      $status->message = "Dashboard sync is disabled.
```
Comment (tlovett1) [inline classes/class-ep-feature.php]: Need to use proper localization functions here.

File: classes/class-ep-dashboard.php [hunk 2]
```diff
@@ -660,7 +716,7 @@ public function action_admin_enqueue_dashboard_scripts() {
 					$index_meta['wpcli_sync'] = true;
 				}
 
-				if ( isset( $_GET['do_sync'] ) ) {
+				if ( isset( $_GET['do_sync'] ) && defined( 'EP_DASHBOARD_SYNC' ) && EP_DASHBOARD_SYNC ) {
```
Comment (tlovett1) [inline classes/class-ep-dashboard.php]: This conditional is wrong.

File: classes/class-ep-feature.php (hunk 1 above)
Comment (tlovett1) [inline classes/class-ep-feature.php]: Oh nevermind.

**Comment by tlovett1** [comment]
Very nice. Just one small tweak needed.

File: classes/class-ep-dashboard.php (hunk 2 above)
Comment (rveitch) [inline classes/class-ep-dashboard.php]: Were you thinking of more strict type checking on the truthiness of EP_DASHBOARD_SYNC? 

Something like this:
```php
if ( isset( $_GET['do_sync'] ) && defined( 'EP_DASHBOARD_SYNC' ) && true === EP_DASHBOARD_SYNC ) {
```

**Comment by tlovett1** [comment]
Can you fix the remaining issue?

**Pull request merged**